src/main/java/org/jenkinsci/plugins/runselector/selectors/PermalinkRunSelector.java

```diff
@@ -41,24 +42,47 @@
 import javax.annotation.Nonnull;
 
 /**
- * Picks up a build through {@link Permalink}
+ * Picks up a build through {@link Permalink}.
  *
  * @author Kohsuke Kawaguchi
  */
 public class PermalinkRunSelector extends AbstractSpecificRunSelector {
-    public final String id;
+
+    @Nonnull
+    private final String id;
 
     @DataBoundConstructor
     public PermalinkRunSelector(String id) {
-        this.id = id;
+        this.id = Util.fixNull(id).trim();
+    }
+
+    @Nonnull
+    public String getId() {
+        return id;
     }
 
     @Override
     @CheckForNull
     public Run<?, ?> getBuild(@Nonnull Job<?, ?> job, @Nonnull RunSelectorContext context) {
-        Permalink p = job.getPermalinks().get(id);
-        if (p==null)    return null;
-        return p.resolve(job);
+        String resolvedId = context.getEnvVars().expand(id);
```

Comment:
+1 for this feature.
This feature allows users to migrate from `SpecificBuildSelector` to `FallbackRunSelector`, `BuildNumberRunSelector` and `PermalinkRunSelector`.